backend/bamboo/blueprints/auth.py

```diff
@@ -1,3 +1,148 @@
-from apiflask import APIBlueprint
+import types
+from datetime import timedelta
+from typing import Any, List
 
+from apiflask import (
+    APIBlueprint,
+    HTTPTokenAuth,
+    abort,
+)
+from flask import current_app
+from jose.exceptions import JWTError
+
+from bamboo.database import db, models
+from bamboo.schemas.auth import LoginSchema, TokenSchema
+from bamboo.utils import decode_jwt, encode_jwt
+
+MANAGE_SITE = 0b001
+MANAGE_USER = 0b010
+MANAGE_CONTENT = 0b100
+PERMISSIONS = (MANAGE_SITE, MANAGE_USER, MANAGE_CONTENT)
+
+
+def auth_required(self, f=None, permissions: int = 0, optional=None) -> Any:
+    """A wrapper of `login_required`.
+
+    Examples:
+    Only login is required.
+    ```python
+    from bamboo.blueprints.auth import token_auth, MANAGE_USER
+
+    @token_auth.auth_required
+    def manage_user_only():
+        ...
+    ```
+
+    One permission is required.
+    ```python
+    from bamboo.blueprints.auth import token_auth, MANAGE_USER
+
+    @token_auth.auth_required(permissions=MANAGE_USER)
+    def manage_user_only():
+        ...
+    ```
+
+    Multiple permissions are required
+    ```python
+    from bamboo.blueprints.auth import (
+        token_auth,
+        MANAGE_SITE,
+        MANAGE_USER,
+    )
+
+    @token_auth.auth_required(permissions=MANAGE_SITE | MANAGE_USER)
+    def manage_site_and_user():
+        ...
+    ```
+    """
+    required_permissions = []
+    for perm in PERMISSIONS:
+        if perm & permissions:
+            required_permissions.append(perm)
+
+    if f:
+        return self.login_required(f=f)
+    return self.login_required(f=None, role=[required_permissions], optional=optional)
```

Comment:
It seems we could override the [`HTTPTokenAuth.authorize()`](https://github.com/miguelgrinberg/Flask-HTTPAuth/blob/52a13b15be17fb058dac160f8c4d460593b6ddce/src/flask_httpauth.py#L115) and implement our own permission checking. Since the original method uses a set comparison, which is less efficient than bitwise check.